### PR TITLE
Add FA3 to SDPA

### PR DESCRIPTION
**Pull request by howardzhang-cv**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #173026
* #172671
* #172622
* __->__ #172040

### Summary: 
Added support for flash attention v3 to SDPA
Only supports fp8 forward pass for now
Created new `_fa3.py` mirroring `_fa4.py` registration
modified `torch/nn/attention/__init__.py` and `_registry.py` to add FA3 support
Added C++ hook to expose FA3 activation (to allow for fp8 dtype)

### Design:
- The design follows the same basic structure as the [FA4 implementation](https://github.com/pytorch/pytorch/pull/167348).
- I added a `_fa3.py` which mirrors the same registration framework as `_fa4.py` (wiring up the new fa3 impl with aten ops `_flash_attention_forward`, `_flash_attention_backward`, `_scaled_dot_product_flash_attention`, `_scaled_dot_product_flash_attention_backward`)
- I additionally added a C++ hook to expose the FA3 activation to `sdp_utils.cpp`, and added a new function to check for low precision dtypes in flash attention (since before fp8 was not allowed)
- Note that only fp8 forward pass is supported for now. We can add fp16/bf16 support later, but fp16/bf16 performance with FA4 is better anyway. The backwards pass currently will throw an error: "FA3 does not support backward pass. Either: 1. Use torch.no_grad() for inference. 2. Unregister FA3 before training: `restore_flash_attention_impl`"

### Test Plan:
Install FA3 library: https://github.com/Dao-AILab/flash-attention/tree/main, follow directions for "FlashAttention-3 beta release", tldr; clone the repo, cd into hopper (where the FA3 implementation lives), python setup.py install

Following the same steps as FA4 to keep it consistent:
```python
activate_flash_attention_impl("FA3")
with sdpa_kernel(SDPBackend.FLASH_ATTENTION):
        _ = F.scaled_dot_product_attention(q, k, v)
```
It also works with torch.compile
```python
activate_flash_attention_impl("FA3")
def sdpa_fn(q, k, v):
      with sdpa_kernel(SDPBackend.FLASH_ATTENTION):
          return F.scaled_dot_product_attention(q, k, v)

compiled_fn = torch.compile(sdpa_fn, fullgraph=True)
_ = compiled_fn(q, k, v)
```

### Some quick runtime results:
I ran some very quick tests with the pytorch profiler to test the runtime of fp8 using FA3 versus bf16 using FA4. Mostly just to catch any easy mistakes slowing it down. These are the results when running on (8, 16, 1024, 128) shape tensors:
FA3 eager:
[sdpa_profiler_fa3_eager.txt](https://github.com/user-attachments/files/24511657/sdpa_profiler_fa3_eager.txt)
FA3 compile:
[sdpa_profiler_fa3_compile.txt](https://github.com/user-attachments/files/24511664/sdpa_profiler_fa3_compile.txt)
FA4 eager:
[sdpa_profiler_fa4_eager.txt](https://github.com/user-attachments/files/24511667/sdpa_profiler_fa4_eager.txt)
FA4 compile:
[sdpa_profiler_fa4_compile.txt](https://github.com/user-attachments/files/24511669/sdpa_profiler_fa4_compile.txt)

### Some quick accuracy results:
I ran a very quick SQNR test on (8, 16, 1024, 128) shape tensors 100 times. I got an average SQNR of 25.55.

This is still a draft, so I'll be doing a quick test with the memory next. Then, I'll do some tests on the diffusers library stable diffusion models.

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -2898,6 +2898,10 @@
   output_differentiability: [True, False, False, False, False, False, False, False, False]
   query, key, value: _scaled_dot_product_flash_attention_backward_symint(grad, query, key, value, output, logsumexp, cum_seq_q, cum_seq_k, max_q, max_k, dropout_p, is_causal, rng_state, unused, scale)
 
+- name: _scaled_dot_product_flash_attention.low_p(Tensor query, Tensor key, Tensor value, Tensor? q_descale, Tensor? k_descale, Tensor? v_descale, float dropout_p=0.0, bool is_causal=False, bool return_debug_mask=False, *, float? scale=None) -> (Tensor output, Tensor logsumexp, Tensor cum_seq_q, Tensor cum_seq_k, SymInt max_q, SymInt max_k, Tensor rng_state, Tensor unused, Tensor debug_attn_mask)
```
Comment (drisspg) [inline tools/autograd/derivatives.yaml]: does the build compile if we dont define a derivative rule for this overload?

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp [hunk 2]
```diff
@@ -678,6 +678,17 @@ bool check_dtypes_low_precision(sdp_params const& params, bool debug) {
   }
 }
 
+bool check_dtypes_flash_attention(sdp_params const& params, bool debug) {
+  auto dprop = at::cuda::getCurrentDeviceProperties();
+  if (dprop->major >= 9 and at::globalContext().userEnabledFA3SDP()) {
+    constexpr auto fa3_dtypes =
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:684]: this should be the union of dtypes right? not just fp8

also is e5m2 supported (although less likely to be used)?

File: torch/nn/attention/_fa3.py [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+    k_descale: torch.Tensor | None,
+    v_descale: torch.Tensor | None,
+) -> str | None:
+    if not all(t.is_cuda for t in tensors):
+        return "inputs must be CUDA tensors"
+    if len({t.device for t in tensors}) != 1:
+        return "inputs must share device"
+    if not all(t.dtype == torch.float8_e4m3fn for t in tensors):
+        return "inputs must be float8_e4m3fn"
+    if q_descale is None or k_descale is None or v_descale is None:
```
Comment (drisspg) [inline torch/nn/attention/_fa3.py]: do you know if fa3 accpets mixed fp8 e.g. fp8 for q@k and diff precision for v?

File: torch/nn/attention/_fa3.py [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+            "When using fp8 SDPA, descale tensor should always be used"
+            " for accurate dequantization. Please use "
+            "F.scaled_dot_product_attention_fp8 and "
+            "provide the descale tensors.",
+            UserWarning,
+        )
+    if q_descale is not None:
+        if q_descale.dtype != torch.float32:
+            return "q_descale must be float32"
+        if not q_descale.is_cuda:
```
Comment (drisspg) [inline torch/nn/attention/_fa3.py]: nit could be cleaned up with a lambda or lil func

File: torch/nn/attention/_fa3.py [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+) -> tuple[torch.Tensor, torch.Tensor]:
+    """
+    Run the FA3 forward pass by calling the C++ kernel directly.
+    """
+    if _FA3_CUDA_FWD is None:
+        raise RuntimeError("FA3 not registered")
+    # Ensure contiguous in the last dimension
+    q = _maybe_contiguous(query)
+    k = _maybe_contiguous(key)
+    v = value.contiguous() if value.stride(-1) != 1 and value.stride(-3) != 1 else value
```
Comment (drisspg) [inline torch/nn/attention/_fa3.py]: I know fa3 does some tricks for v majornes on hopper is that what this is doing?

File: torch/nn/attention/_fa3.py [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+        q_descale,  # q_descale,
+        k_descale,  # k_descale,
+        v_descale,  # v_descale,
+        scale,  # softmax_scale,
+        is_causal,  # causal,
+        window_size_left if window_size_left is not None else -1,  # window_size_left
+        window_size_right if window_size_right is not None else -1,  # window_size_right
+        0,  # attention_chunk,
+        0.0,  # softcap,
+        True,  # rotary_interleaved,
```
Comment (drisspg) [inline torch/nn/attention/_fa3.py:312]: should this be set?

File: torch/nn/attention/_fa3.py [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+        None,  # scheduler_metadata,
+        1,  # num_splits,
```
Comment (drisspg) [inline torch/nn/attention/_fa3.py]: 1 is auto handle and then we typically force this to non split w/ determinism is set: I have a pr doing this for fa4

File: torch/nn/attention/_fa3.py [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+        None,  # pack_gqa,
+        0,  # sm_margin,
```
Comment (drisspg) [inline torch/nn/attention/_fa3.py]: we techincally have this thing called sm_carveout that should be wired up for this

File: torch/nn/attention/_fa3.py [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+    q_descale: torch.Tensor | None = None,
+    k_descale: torch.Tensor | None = None,
+    v_descale: torch.Tensor | None = None,
+    *,
+    scale: float | None = None,
+    window_size_left: int = -1,
+    window_size_right: int = -1,
+):
+    raise RuntimeError(
+        "FA3 does not support backward pass. Either:\n"
```
Comment (drisspg) [inline torch/nn/attention/_fa3.py]: I think we should wire this up just error for low p for now

File: torch/nn/functional.py [hunk 10]
```diff
@@ -6119,6 +6119,69 @@ def forward(self, ...):
 )
 
 
+def scaled_dot_product_attention_fp8(
```
Comment (drisspg) [inline torch/nn/functional.py]: talked offline we should have this be private for now or non exposed while we work on API and can add to ao

File: torch/nn/functional.pyi.in [hunk 11]
```diff
@@ -896,6 +896,19 @@ def upsample(
 
 __all__ += ["upsample"]
 
+def scaled_dot_product_attention_fp8(
```
Comment (drisspg) [inline torch/nn/functional.pyi.in]: above comment

File: torchgen/aoti/fallback_ops.py [hunk 12]
```diff
@@ -36,7 +36,9 @@
     "aten._fft_c2c.default": {},
     "aten._fft_r2c.default": {},
     "aten._flash_attention_backward.default": {},
+    "aten._flash_attention_backward.low_p": {},
```
Comment (drisspg) [inline torchgen/aoti/fallback_ops.py]: we shud bikeshed name w/ AO folks

File: aten/src/ATen/native/native_functions.yaml [hunk 13]
```diff
@@ -15232,12 +15243,24 @@
...
 - func: _flash_attention_backward(Tensor grad_out, Tensor query, Tensor key, Tensor value, Tensor out, Tensor logsumexp, Tensor cum_seq_q, Tensor cum_seq_k, SymInt max_q, SymInt max_k, float dropout_p, bool is_causal, Tensor rng_state, Tensor unused, *, float? scale=None, SymInt? window_size_left=None, SymInt? window_size_right=None) -> (Tensor, Tensor, Tensor)
   device_check: NoCheck
   variants: function
   dispatch:
     CUDA: _flash_attention_backward
 
+- func: _flash_attention_backward.low_p(Tensor grad_out, Tensor query, Tensor key, Tensor value, Tensor out, Tensor logsumexp, Tensor cum_seq_q, Tensor cum_seq_k, SymInt max_q, SymInt max_k, float dropout_p, bool is_causal, Tensor rng_state, Tensor unused, Tensor? q_descale, Tensor? k_descale, Tensor? v_descale, *, float? scale=None, SymInt? window_size_left=None, SymInt? window_size_right=None) -> (Tensor, Tensor, Tensor)
+  device_check: NoCheck
+  variants: function
+  dispatch:
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml]: I think just remove the bwd low p variants for now

File: torch/nn/attention/_fa3.py (hunk 3 above)
Comment (howardzhang-cv) [reply torch/nn/attention/_fa3.py]: I looked through and it seems like it does not. It only supports same precision for fp8 forward, and fp16 and bf16 forward + backward

File: torch/nn/attention/_fa3.py (hunk 5 above)
Comment (howardzhang-cv) [reply torch/nn/attention/_fa3.py]: Yeah it's a separate check because V can be column major or row major, but this is only for fp8 so adding an extra check now.

File: torch/nn/attention/_fa3.py (hunk 6 above)
Comment (howardzhang-cv) [reply torch/nn/attention/_fa3.py:312]: In the _flash_attn_forward interface in hopper/flash_attn_interface.py it seems to default to True so I put True here

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (howardzhang-cv) [reply tools/autograd/derivatives.yaml]: It does compile, just removed the derivative rule for the overload.

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp (hunk 2 above)
Comment (howardzhang-cv) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:684]: Looked into flash-attention repo and found that e5m2 is not supported

File: torch/nn/attention/_fa3.py (hunk 4 above)
Comment (howardzhang-cv) [reply torch/nn/attention/_fa3.py]: Changed

File: torch/nn/attention/_fa3.py (hunk 7 above)
Comment (howardzhang-cv) [reply torch/nn/attention/_fa3.py]: Added the check for determinism and set to be auto instead of always 1. Can you look through and see if this is correct?

File: torch/nn/attention/_fa3.py (hunk 8 above)
Comment (howardzhang-cv) [reply torch/nn/attention/_fa3.py]: Added sm_carveout check

File: torch/nn/attention/_fa3.py (hunk 9 above)
Comment (howardzhang-cv) [reply torch/nn/attention/_fa3.py]: Wired up fp16 and bf16 backwards pass, and throw error if fp8

File: torch/nn/functional.py (hunk 10 above)
Comment (howardzhang-cv) [reply torch/nn/functional.py]: I added the "_" in the beginning and removed it from the functional.pyi.in so it doesn't get caught by wildcard imports and things like that, is this enough, or should we maybe move this into an experimental folder somewhere?

File: aten/src/ATen/native/native_functions.yaml (hunk 13 above)
Comment (howardzhang-cv) [reply aten/src/ATen/native/native_functions.yaml]: Removed

**Comment by drisspg** [comment]
@claude can you review this PR please

**Comment by drisspg** [comment]
@claude review this pr please

File: torch/nn/attention/_fa3.py (hunk 7 above)
Comment (drisspg) [reply torch/nn/attention/_fa3.py]: looks good

File: torch/nn/functional.py (hunk 10 above)
Comment (drisspg) [reply torch/nn/functional.py]: I think we should probably put this into a _prototype like folder in the attention repo and not int he main nn.functional just yet

**Review by drisspg — APPROVED** [review]
